internal/repository/ent/priceunit.go

```diff
@@ -0,0 +1,290 @@
+package ent
+
+import (
+	"context"
+
+	"github.com/flexprice/flexprice/ent"
+	"github.com/flexprice/flexprice/ent/priceunit"
+	"github.com/flexprice/flexprice/internal/cache"
+	domainPriceUnit "github.com/flexprice/flexprice/internal/domain/priceunit"
+	ierr "github.com/flexprice/flexprice/internal/errors"
+	"github.com/flexprice/flexprice/internal/logger"
+	"github.com/flexprice/flexprice/internal/postgres"
+	"github.com/flexprice/flexprice/internal/types"
+	"github.com/shopspring/decimal"
+)
+
+type priceUnitRepository struct {
+	client postgres.IClient
+	log    *logger.Logger
+	cache  cache.Cache
+}
+
+// NewPriceUnitRepository creates a new instance of priceUnitRepository
+func NewPriceUnitRepository(client postgres.IClient, log *logger.Logger, cache cache.Cache) domainPriceUnit.Repository {
+	return &priceUnitRepository{
+		client: client,
+		log:    log,
+		cache:  cache,
+	}
+}
+
+func (r *priceUnitRepository) Create(ctx context.Context, unit *domainPriceUnit.PriceUnit) error {
+	client := r.client.Querier(ctx)
+
+	_, err := client.PriceUnit.Create().
+		SetID(unit.ID).
+		SetName(unit.Name).
+		SetCode(unit.Code).
+		SetSymbol(unit.Symbol).
+		SetBaseCurrency(unit.BaseCurrency).
+		SetConversionRate(unit.ConversionRate).
+		SetPrecision(unit.Precision).
+		SetStatus(string(types.StatusPublished)). // Set default status to published
+		SetTenantID(unit.TenantID).
+		SetEnvironmentID(unit.EnvironmentID).
+		SetCreatedAt(unit.CreatedAt).
+		SetUpdatedAt(unit.UpdatedAt).
+		Save(ctx)
+
+	if err != nil {
+		if ent.IsConstraintError(err) {
+			return ierr.WithError(err).
+				WithHint("A pricing unit with this code already exists").
+				WithReportableDetails(map[string]any{
+					"code": unit.Code,
+				}).
+				Mark(ierr.ErrAlreadyExists)
+		}
+		return ierr.WithError(err).
+			WithHint("Failed to create pricing unit").
+			Mark(ierr.ErrDatabase)
+	}
+
+	return nil
+}
+
+// GetByID retrieves a pricing unit by ID
+func (r *priceUnitRepository) GetByID(ctx context.Context, id string) (*domainPriceUnit.PriceUnit, error) {
+	client := r.client.Querier(ctx)
+
+	unit, err := client.PriceUnit.Get(ctx, id)
+	if err != nil {
+		if ent.IsNotFound(err) {
+			return nil, ierr.WithError(err).
+				WithHint("Pricing unit not found").
+				Mark(ierr.ErrNotFound)
+		}
+		return nil, ierr.WithError(err).
+			WithHint("Failed to get pricing unit").
+			Mark(ierr.ErrDatabase)
+	}
+
+	return domainPriceUnit.FromEnt(unit), nil
+}
+
+func (r *priceUnitRepository) List(ctx context.Context, filter *domainPriceUnit.PriceUnitFilter) ([]*domainPriceUnit.PriceUnit, error) {
```

Comment:
refer count and list method of plans/prices and see how we have reued queryies and played with filter object